File: herencia-dto/SergioDeIscar-herencia-java/SergioDeIscar-herencia-java/src/main/java/org/example/factories/PersonaFactory.java
package org.example.factories;

import org.example.models.Alumno;
import org.example.models.Persona;
import org.example.models.Profesor;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class PersonaFactory {
    //Singleton
    private static PersonaFactory instance = null;
    public static PersonaFactory getInstance() {
        if (instance == null) {
            instance = new PersonaFactory();
        }
        return instance;
    }

    public List<Persona> getRdnPersonas(){
        List<Persona> personas = new ArrayList<>();
        for (int i = 0; i < 20; i++) {
            personas.add(
                getRdnPersona()
            );
        }
        return personas;
    }

    public Persona getRdnPersona(){
        int rdn = (int) (Math.random() * 100);
        Random random = new Random();
        if (rdn < 10) {
            return new Alumno(
                getRdnNombre(),
                (random.nextInt(40-18) + 18)
            );
        } else {
            return new Profesor(
                getRdnNombre(),
                getRdnModulo()
            );
        }
    }

    private String getRdnNombre(){
        String[] nombres = {"Pepe", "Juan", "Ana"};
        return nombres[(int) (Math.random() * nombres.length)];
    }
    private String getRdnModulo(){
        String[] modulos = {"Programación", "Entornos"};
        return modulos[(int) (Math.random() * modulos.length)];
    }
}

File: herencia-dto/SergioDeIscar-herencia-kotlin/SergioDeIscar-herencia-kotlin/src/main/kotlin/factories/PersonaFactory.kt
package factories

import models.Alumno
import models.Persona
import models.Profesor

object PersonaFactory {
    fun getRdnPersonas(): List<Persona>{
        val list = mutableListOf<Persona>()
        repeat(20){ list.add(getRdnPersona()) }
        return list
    }

    fun getRdnPersona(): Persona{
        return when((0..100).random()){
            in 0..10 -> Alumno(
                nombre = getRdnNombre(),
                edad = (18..40).random()
            )
            else -> Profesor(
                nombre = getRdnNombre(),
                modulo = getRdnModulo()
            )
        }
    }

    private fun getRdnNombre(): String{
        return arrayOf("Pepe", "Juan", "Ana").random()
    }
    private fun getRdnModulo(): String{
        return arrayOf("Programación", "Entornos").random()
    }

    fun getPersonasDefault(): List<Persona> = listOf(
        Alumno(1, "Pepe", 20),
        Profesor(2, "Juan", "Programación"),
        Alumno(3, "Ana", 18),
        Profesor(4, "Pepe", "Entornos")
    )
}